File: 0438-find-all-anagrams-in-a-string/python/Solution.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        n = len(s)
        m = len(p)
        if m > n:
            return []
        needs = [0] * 26
        for i in range(m):
            needs[ord(p[i]) - ord("a")] +=1
        matched = [0] * 26
        startp = 0
        endp = 0
        result = []
        while endp < m:
            matched[ord(s[endp]) - ord("a")] += 1
            endp += 1
        if self.same(needs, matched):
            result.append(startp)
        while endp < n and startp < n:
            matched[ord(s[startp]) - ord("a")] -= 1
            matched[ord(s[endp]) - ord("a")] += 1
            startp += 1
            endp += 1
            if self.same(needs, matched):
                result.append(startp)
        return result

    def same(self, needs, matched):
        for i in range(len(needs)):
            if needs[i] != matched[i]:
                return False
        return True

File: 0438-find-all-anagrams-in-a-string/cpp/Solution.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        int n = s.size();
        int m = p.size();
        if (m > n) return {};
        vector<int> needs(26);
        for (int i = 0; i < m; ++i) {
            needs[p[i] - 'a']++;
        }
        vector<int> matched(26);
        int startp = 0;
        int endp = 0;
        vector<int> result;
        while (endp < m) {
            matched[s[endp] - 'a']++;
            endp++;
        }
        if (same(needs, matched)) {
            result.push_back(startp);
        }
        while (endp < n && startp < n) {
            matched[s[startp] - 'a']--;
            matched[s[endp] - 'a']++;
            startp++;
            endp++;
            if (same(needs, matched)) {
                result.push_back(startp);
            }
        }
        return result;
    }
    bool same(const vector<int>& needs, const vector<int>& matched) {
        for (int i = 0; i < needs.size(); ++i) {
            if (needs[i] != matched[i]) return false;
        }
        return true;
    }
};